### PR TITLE
fix: stop using Arc::ptr_eq with dyn Trait

### PR DESCRIPTION
**Pull request by koivunej**
This changes the way we compare `Arc<dyn PersistentLayer>` in Timeline's `LayerMap` not to use `Arc::ptr_eq` which has been witnessed in development of #3557 to yield wrong results. It gives wrong results because it compares fat pointers, which are `(object, vtable)` tuples for `dyn Trait` and there are no guarantees that the `vtable`s are unique. As in there were multiple vtables for `RemoteLayer` which is why the comparison failed in #3557.

This is a known issue in rust, clippy warns against it and rust std might be moving to the solution which has been reproduced on this PR: compare only object pointers by "casting out" the vtable pointer.

### CODE REVIEW

**Review by problame — APPROVED** [review]
It solved my problems in #3557

File: pageserver/src/tenant/layer_map.rs [hunk 1]
```diff
@@ -733,14 +733,27 @@ where
...
+        // A future version of rust will most likely use this form below, where we cast each
+        // pointer into a pointer to unit, which drops the inaccessible vtable pointer, making it
+        // not affect the comparison.
+        //
+        // See: https://github.com/rust-lang/rust/pull/106450
+        let left = Arc::as_ptr(left) as *const ();
+        let right = Arc::as_ptr(right) as *const ();
+
         // In future rust versions this might become Arc::as_ptr(left) as *const () ==
         // Arc::as_ptr(right) as *const (), we could change to that before.
```
Comment (koivunej) [inline pageserver/src/tenant/layer_map.rs:755]:
```suggestion
```

I think the comments above are enough :)

**Comment by problame** [comment]
`regress-test (debug)` failed, suspecting spurious failure.
https://github.com/neondatabase/neon/actions/runs/4115585724/attempts/1

Kicking it for a re-run.

**Comment by bojanserafimov** [comment]
Are we going through all this effort because layers don't have `Hash` and `Eq`? We don't use this method on the read path, so why not just implement those methods based on layer name?

**Comment by knizhnik** [comment]
> Are we going through all this effort because layers don't have Hash and Eq? We don't use this method on the read path, so why not just implement those methods based on layer name?

Calculating hash and performing comparison of pointer is much faster than for layer names (more than 100 characters)

**Comment by koivunej** [comment]
> Are we going through all this effort because layers don't have `Hash` and `Eq`? We don't use this method on the read path, so why not just implement those methods based on layer name?

As for Layer's identity being the pointer even though we don't have an explicit file => `Arc<Layer>` interning registry, comparing Arc pointers seems valid, and is only 8 bytes, doesn't involve any downcasting, so it is perhaps the cheapest way. Even if we had the "less dyn" type refactorings which @SomeoneToIgnore has proposed many times, we would probably still want to keep the "Layers identity is the Arc pointer".

Why we ran into this now was that #3557 is that perhaps enough code was added to force creation of two distinct vtable pointers for RemoteLayer. Tested the branch in both --release and dev profiles for the same outcome.

**Comment by koivunej** [comment]
I think this PR is ready, but would appreaciate more reviews.

This would be good to include if we ever do any hotfix for the current release.

**Review by hlinnaka — APPROVED** [review]
The comment makes it sound like this is a problem particularly with rust 1.66.1, but as far as I understand it's the same on all versions. Maybe rephrase it as:

> as of this writing (at least up to rust 1.67), `Arc::ptr_eq` uses a comparison method ...

Other than that, looks good to me

File: pageserver/src/tenant/layer_map.rs [hunk 2]
```diff
@@ -733,14 +733,25 @@ where
...
-        // Arc::as_ptr(right) as *const (), we could change to that before.
-        #[allow(clippy::vtable_address_comparisons)]
-        Arc::ptr_eq(left, right)
+        // rust does not provide a guarantee that these vtable's are unique, but however
+        // `Arc::ptr_eq` in 1.66.1 uses a comparison where both the pointer and the vtable need to
+        // be equal.
+        //
+        // See: https://github.com/rust-lang/rust/issues/103763
+        //
+        // A future version of rust will most likely use this form below, where we cast each
```
Comment (knizhnik) [inline pageserver/src/tenant/layer_map.rs:746]: Sorry, I do not understand this comment and why we need to case pointer to "const*?
Why it is not possible just to do Arc::as_ptr(left) == Arc::as_ptr(right) as suggested in Rust doc:
```
/// Provides a raw pointer to the data.
    ///
    /// The counts are not affected in any way and the `Arc` is not consumed. The pointer is valid for
    /// as long as there are strong counts in the `Arc`.
    ///
    /// # Examples
    ///
    /// ```
    /// use std::sync::Arc;
    ///
    /// let x = Arc::new("hello".to_owned());
    /// let y = Arc::clone(&x);
    /// let x_ptr = Arc::as_ptr(&x);
    /// assert_eq!(x_ptr, Arc::as_ptr(&y));
    /// assert_eq!(unsafe { &*x_ptr }, "hello");
    /// ```
    #[must_use]
    #[stable(feature = "rc_as_ptr", since = "1.45.0")]
    pub fn as_ptr(this: &Self) -> *const T {
        let ptr: *mut ArcInner<T> = NonNull::as_ptr(this.ptr);

        // SAFETY: This cannot go through Deref::deref or RcBoxPtr::inner because
        // this is required to retain raw/mut provenance such that e.g. `get_mut` can
        // write through the pointer after the Rc is recovered through `from_raw`.
        unsafe { ptr::addr_of_mut!((*ptr).data) }
    }
```

Comment (koivunej) [reply pageserver/src/tenant/layer_map.rs:746]: I'll try to come up with playground link to illustrate this.

Comment (koivunej) [reply pageserver/src/tenant/layer_map.rs:746]: Does this illustrate the two-pointer nature of `*const dyn Trait` as given by `Arc::as_ptr` in this case?

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0d1038e802b245e2f29d9264b8b077d8

Comment (koivunej) [reply pageserver/src/tenant/layer_map.rs:746]: Also compare to `Arc::ptr_eq` impl: https://doc.rust-lang.org/stable/src/alloc/sync.rs.html#1150

So this is what we are hoping to do differently, until std adds a different method or changes that `Arc::ptr_eq` method, as has been voted on the linked issue.

Comment (koivunej) [reply pageserver/src/tenant/layer_map.rs:746]: Also compare to `Arc::ptr_eq` impl: https://doc.rust-lang.org/stable/src/alloc/sync.rs.html#1150

So this is what we are hoping to do differently, until std adds a different method or changes that `Arc::ptr_eq` method, as has been voted on the linked issue.

**Comment by koivunej** [comment]
Transient `test_runner/regress/test_tenant_size.py::test_get_tenant_size_with_multiple_branches` after the doc update. Will report it to the issue and re-run.

**Review by knizhnik — APPROVED** [review]
